extensions/2.0/Vendor/NV_materials_mdl/README.md

```diff
@@ -0,0 +1,511 @@
+        },
+        "value": "some string value"
+    },
+    {
+        "name": "arg_float2x3_array",
+        "type": {
+            "typeName": "float2x3",
+            "arraySize": 2
+        },
+        "value": [
+            [ 1.0, 0.0, 0.0, 1.0, 0.0, 0.0 ],
+            [ 1.0, 0.0, 0.0, 1.0, 0.0, 0.0 ]
+        ]
+    },
+    {
+        "name": "arg_enum",
+        "type": {
+            "typeName": "intensity_mode"
+        },
+        "value": "intensity_radiant_exitance"
+    }
+]
+```
+
+## Special Functions
+A conforming implementation is required to recognize a set of special functions.
+
+### Array Constructor
+The array constructor function has the special name `T[]` and takes `N` arguments with the array element type `Type` named `value0`, `value1`, `value2`, and so on. The return type must be the array element type `Type` and array size `N`.
+
+Example:
+```json
+{
+    "functionName": "T[]",
+    "type": {
+        "typeName": "float",
+        "arraySize": 3
+    },
+    "arguments": [
+        {
+            "name": "value0",
+            "type": {
+                "typeName": "float"
+            },
+            "value": 42.0
+        },
+        {
+            "name": "value1",
+            "functionCall": 1
+        },
+        {
+            "name": "value2",
+            "functionCall": 2
+        }
+    ]
+}
+```
+
+### Member Accessor
+Member accessors (the *dot operator*) are named after the member being accessed preceded by the structures name followed by the dot operator (e.g. `texture_return.tint` or `float3.x`). The function has a single argument `s` which is the instance of the structure type from which the member should be accessed and the return type is the type of the member. Unlike most other special functions the member accessor is defined in the module in which the structure type is defined.
+
+Example:
+```json
+{
+    "module": 1,
+    "functionName": "texture_return.tint",
+    "type": {
+        "typeName": "color"
+    },
+    "arguments": [
+        {
+            "name": "s",
+            "functionCall": 2
+        }
+    ]
+}
+```
+
+### Type-Cast Operator
+The name of the type-cast operator function is `operator_cast` and has a single argument `cast` with the type `Type`. The return type is a cast-compatible type `Type2` as defined in the [MDL Language Specification](https://github.com/NVIDIA/MDL-SDK/tree/master/doc/specification).
+
+Example:
+```json
+{
+    "functionName": "operator_cast",
+    "type": {
+        "module": 0,
+        "typeName": "two_colors"
+    },
+    "arguments": [
+        {
+            "name": "cast",
+            "type": {
+                "module": 0,
+                "typeName": "color_pair"
+            },
+            "functionCall": 1
+        }
+    ]
+}
+```
+
+### Array Length Operator
+The name of the array length operator function is `operator_len` and has a single argument `a` which must be an array type. The return type is `int`.
```

Comment:
Given that the number of array elements is known at compile time, why is this operator needed?

Reply:
This operator is useful for composing function call graphs in authoring tools. Even if the array size is known, querying it makes it more convenient in some cases when authoring the graphs. For completeness, we want to be able to represent all possible expressions in the call graph. We don't want to require having different call graphs during import and export, due to replacing function calls to the array size operator with constants. This choice should be left to the authoring tools.
Granted, the example for this operator in the README could have easily been replaced with a constant. I changed the example, so it calls the operator on the return value of a function call, which is a more realistic use-case.